src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -449,17 +449,145 @@ private bool TryCreateCountSegment(string segmentText)
+        /// <param name="filter">Filter expression.</param>
+        /// <returns>Filter clause with the navigation source and filter information.</returns>
+        private FilterClause GenerateFilterClause(IEdmNavigationSource navigationSource, string filter)
+        {
+            Debug.Assert(navigationSource != null, "navigationSource != null");
+            Debug.Assert(filter.Length > 0, "filter.Length > 0");
+
+            IEdmEntitySet entitySet = navigationSource as IEdmEntitySet;
+            ODataPathInfo currentOdataPathInfo = new ODataPathInfo(entitySet.EntityType(), entitySet);
+
+            // Get the syntactic representation of the filter expression.
+            UriQueryExpressionParser expressionParser = new UriQueryExpressionParser(
+                configuration.Settings.FilterLimit, configuration.EnableCaseInsensitiveUriFunctionIdentifier);
+            QueryToken filterToken = expressionParser.ParseFilter(filter);
+
+            // Bind it to metadata.
+            BindingState state = new BindingState(configuration, currentOdataPathInfo.Segments.ToList())
+            {
+                ImplicitRangeVariable = NodeFactory.CreateImplicitRangeVariable(
+                currentOdataPathInfo.TargetEdmType.ToTypeReference(), currentOdataPathInfo.TargetNavigationSource)
+            };
+            state.RangeVariables.Push(state.ImplicitRangeVariable);
+
+            MetadataBinder binder = new MetadataBinder(state);
+            FilterBinder filterBinder = new FilterBinder(binder.Bind, state);
+
+            return filterBinder.BindFilter(filterToken);
+        }
+
+        /// <summary>
+        /// Try to handle the segment as $filter.
+        /// </summary>
+        /// <param name="segmentText">The raw segment text.</param>
+        /// <returns>Whether the segment was $filter.</returns>
+        /// <remarks>$filter path segment is different from existing path segments in that it strictly
+        /// follows the format of "$filter=@a", where @a represents a parameter alias. Thus, this function
+        /// should validate the format of the path segment closely.</remarks>
+        private bool TryCreateFilterSegment(string segmentText)
+        {
+            Debug.Assert(segmentText != null, "segmentText != null");
+
+            /*
+             * 1) Check whether the path segment starts with $filter.
+             * 2) Ensure that "=@" that follows the identifier and that there's only one '@'.
+             * 3) Extract the parameter alias and validate it syntactically.
+             * 4) Add the filter segment to list of parsed segments.
+             */
+
+            // 1) Check whether the path segment starts with $filter. Past this point, we will throw invalid syntax exceptions
+            // because we will assume that the user is attempting to use the $filter path segment.
+            if (!segmentText.StartsWith(UriQueryConstants.FilterSegment, StringComparison.OrdinalIgnoreCase))

```

Comment:
Use EnableCaseInsensitiveUriFunctionIdentifier (which needs to be renamed to remove "UriFunctionIdentifier").